# Primeira parte

## A. artists.csv

### - Obtendo a dimensao do conjunto de dados de "artists"

In [3]:
import pandas as pd

# Carregando o arquivo da planilha
artists_df = pd.read_csv('/home/hjhgjhgjhg/Downloads/csv/artists.csv')

# Obtendo as dimensões da planilha
dimensions = artists_df.shape
print("Dimensao do conjunto de dados do csv de artists:", dimensions)

# Imprimindo as dimensões da planilha
print(f"Ou seja: {dimensions[0]} linhas e {dimensions[1]} colunas")

Dimensao do conjunto de dados do csv de artists: (307, 6)
Ou seja: 307 linhas e 6 colunas


/tmp/ipykernel_4675/2518351264.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


###  - O nome das colunas e que tipo de dados que lhe estão associados

In [4]:
# Carregue o arquivo da planilha
artists_df = pd.read_csv('/home/hjhgjhgjhg/Downloads/csv/artists.csv')

# Obtenha o nome das colunas e seus tipos de dados
column_names = artists_df.columns.tolist()
column_types = artists_df.dtypes.tolist()

# Imprima o nome das colunas e seus tipos de dados
for i in range(len(column_names)):
    print(f"A coluna '{column_names[i]}' é do tipo de dados '{column_types[i]}'")

A coluna 'artist' é do tipo de dados 'object'
A coluna 'collab_songs' é do tipo de dados 'int64'
A coluna 'type' é do tipo de dados 'object'
A coluna 'genres' é do tipo de dados 'object'
A coluna 'collab_individuals' é do tipo de dados 'int64'
A coluna 'country' é do tipo de dados 'object'


#### É possível notar que o elemento em comum entre todos os dados de cada coluna foi obtido a partir do tipo dela, pois cada coluna possui todos os seus dados de um mesmo tipo, tendo um valor único

###  - Existem valores NaN? Se sim, em quantidade significativa? Em que colunas? 

In [ ]:
print("\nQuantidade de valores NaN por coluna:")

# Contando a quantidade de valores NaN em cada coluna
nan_count = artists_df.isna().sum()

# Imprimindo a quantidade de valores NaN em cada coluna
for i in range(len(nan_count)):
    print(f"A coluna '{nan_count.index[i]}' contém {nan_count[i]} valores NaN")


Quantidade de valores NaN por coluna:
A coluna 'artist' contém 0 valores NaN
A coluna 'collab_songs' contém 0 valores NaN
A coluna 'type' contém 0 valores NaN
A coluna 'genres' contém 0 valores NaN
A coluna 'collab_individuals' contém 0 valores NaN
A coluna 'country' contém 0 valores NaN


#### É notório que não. Não existem valores não numéricos nesse arquivo csv em nenhuma das colunas.

### Explore se o dataset contém algum artista de “Fado”, um tipo de música tradicional de Portugal. 

In [12]:
# Coluna a contar os dados
column_name = 'genres'

# Quantidade de valores iguais na coluna
value_counts = artists_df[column_name].value_counts()

# Imprimindo a quantidade de valores iguais na coluna
print(f"A coluna '{column_name}' contém os seguintes valores e suas respectivas quantidades:\n{value_counts}")

A coluna 'genres' contém os seguintes valores e suas respectivas quantidades:
genres
Sense dades                                                                    76
Latin, Reggaeton                                                               25
Hip hop                                                                        11
Hip hop, Trap                                                                   5
Hip hop, Latin, Reggaeton                                                       4
                                                                               ..
Dance, Edm, Electronic, House, Trance                                           1
Dance-pop, Indie pop, Pop                                                       1
Afrobeats, Dancehall, Reggae                                                    1
Hardcore hip hop, Hip hop, Horrorcore, Pop rap                                  1
Avant-garde jazz, Free improvisation, Free jazz, Jazz, Post-bop, Vocal jazz     1
Name: count, 

#### Ao analisar os dados afquiridos na tabela, podemos ver que não existe nenhum cantor na tabela que que cante o tipo de música "fado".

### - Qual é o número máximo de colaborações de um artista neste dataset? Identifique esse artista.

In [ ]:
# Coluna 1 referente ao artista e a coluna 2 referente a quantidade de collab songs
column_1 = 'artist'
column_2 = 'collab_songs'

# Obtendo o objeto da coluna 1 com o maior valor na coluna 2
max_value = artists_df[column_2].max()
max_value_row = artists_df.loc[artists_df[column_2] == max_value]

max_value_object = max_value_row[column_1].iloc[0]

# Imprimindo o artista da coluna 1 com o maior quantidade de collab songs na coluna 2
print(f"O artista com o maior valor na de colaborações em músicas (collab songs) é: {max_value_object}")
print(f"{max_value_object} possui {max_value} collab songs")

O artista com o maior valor na de colaborações em músicas (collab songs) é: Lil Wayne
Lil Wayne possui 525 collab songs


In [ ]:
# Coluna 1 referente ao artista e a coluna 2 referente a quantidade de collab songs
column_1 = 'artist'
column_2 = 'collab_individuals'

# Obtendo o objeto da coluna 1 com o maior valor na coluna 2
max_value = artists_df[column_2].max()
max_value_row = artists_df.loc[artists_df[column_2] == max_value]

max_value_object = max_value_row[column_1].iloc[0]

# Imprimindo o artista da coluna 1 com o maior quantidade de collab songs na coluna 2
print(f"O artista com o maior valor na de colaborações individuais (collab individuals) é: {max_value_object}")
print(f"{max_value_object} possui {max_value} collab individuals")

O artista com o maior valor na de colaborações individuais (collab individuals) é: Lil Wayne
Lil Wayne possui 65 collab individuals


#### Com isso, é notório que o artista que mais contribuiu, tanto em collab songs quanto collab individuals foi o Lil Wayne com 525 collab songs e 65 collab individuals.

### - Existem artistas do Brasil neste dataset? Se sim, liste a informação a seu respeito, apresentando-a por ordem decrescente de músicas criadas em colaboração. 

In [16]:
# Filtra os artistas do Brasil no DataFrame artists_df
brazilian_artists = artists_df[artists_df['country'] == 'Brazil']

# Ordena os artistas brasileiros pelo número de colaborações em ordem decrescente
brazilian_artists_sorted = brazilian_artists.sort_values(by='collab_songs', ascending=False)

# Imprime os artistas do Brasil no dataset, ordenados por colaborações
print("Artistas do Brasil no dataset (ordenados por colaborações):")
print(brazilian_artists_sorted)


Artistas do Brasil no dataset (ordenados por colaborações):
                  artist  collab_songs   type  \
249               Anitta            52  Woman   
12      Marília Mendonça            46  Woman   
11      Maiara & Maraisa            44   Band   
48           Luísa Sonza            16  Woman   
269       Jorge & Mateus            12   Band   
145        Gusttavo Lima            11    Man   
142       Wesley Safadão            11    Man   
20             Zé Felipe            10    Man   
141             Vulgo FK            10    Man   
1           Luan Santana             9    Man   
192     Hugo & Guilherme             8   Band   
40                L7NNON             7    Man   
75                Chefin             7    Man   
61        Mari Fernandez             6  Woman   
35              DJ Guuga             5    Man   
299        PEDRO SAMPAIO             4    Man   
74     Marcynho Sensação             4    Man   
157        Felipe Amorim             3    Man   
237      

### Quais são os 3 países listados no dataset que apresentam maior número de colaborações? E
### quais são os 3 países que apresentam menor número de colaborações.

In [1]:
top_countries_collaborations = artists_df.groupby('country')['collab_songs'].sum().nlargest(4)

# Removendo a entrada correspondente a "Outros países"
top_countries_collaborations = top_countries_collaborations[top_countries_collaborations.index != 'Other countries']

print("Top 3 países com maior número de colaborações:")
print(top_countries_collaborations.head(3))

bottom_countries_collaborations = artists_df.groupby('country')['collab_songs'].sum().nsmallest(3)
print("\nTop 3 países com menor número de colaborações:")
print(bottom_countries_collaborations)


NameError: name 'artists_df' is not defined

Os Estados Unidos lideram o ranking de colaborações musicais com 9481 registros, refletindo sua influência global na indústria musical. Em segundo lugar, Porto Rico apresenta 1777 colaborações, destacando sua contribuição significativa, especialmente nos gêneros latinos e reggaeton. O Canadá ocupa o terceiro lugar com 664 colaborações, refletindo sua diversificada cena musical e contribuição para a música global. Por outro lado, a Coreia do Sul, Nigéria e Itália figuram entre os países com menor número de colaborações, com 69, 74 e 80 registros, respectivamente, indicando uma presença menos proeminente na indústria musical global.

### Como se ordenam os continentes nesta lista, por número de artistas representados?

In [17]:
import pandas as pd

# Dados fornecidos
data = {
    "artist": ["Arcángel", "Luan Santana", "50 Cent", "Nicki Nicole", "María Becerra", "Ana Castela", "Stromae", "Jason Derulo", "Farruko", "Camila Cabello", "Cris Mj", "Maiara & Maraisa", "Marília Mendonça", "Jung Kook", "De La Ghetto", "SZA", "Zion & Lennox", "Tyga", "blackbear", "Pusha T"],
    "collab_songs": [163, 9, 241, 19, 17, 2, 5, 49, 150, 35, 1, 44, 46, 11, 78, 46, 80, 198, 22, 98],
    "type": ["Man", "Man", "Man", "Woman", "Woman", "Woman", "Man", "Man", "Man", "Woman", "Man", "Band", "Woman", "Man", "Man", "Woman", "Band", "Man", "Man", "Man"],
    "country": ["Puerto Rico", "Brazil", "United States", "Argentina", "Argentina", "Brazil", "Other countries", "United States", "Puerto Rico", "United States", "Other countries", "Brazil", "Brazil", "South Korea", "Puerto Rico", "United States", "Puerto Rico", "United States", "United States", "United States"],
}

artists_df = pd.read_csv('/home/hjhgjhgjhg/Downloads/csv/artists.csv')

# Mapeando os países para continentes
country_to_continent = {
    'Puerto Rico': 'North America',
    'Brazil': 'South America',
    'United States': 'North America',
    'Argentina': 'South America',
    'South Korea': 'Asia',
    'Outros países': 'Outros'
}

# Função para mapear o continente com base no país
def map_country_to_continent(country):
    return country_to_continent.get(country, 'Não mencionado')

# Aplicando a função de mapeamento para obter os continentes
artists_df['continent'] = artists_df['country'].apply(map_country_to_continent)

# Contando quantos artistas há por continente
artists_by_continent = artists_df['continent'].value_counts()

print(artists_by_continent)



continent
North America     153
Não mencionado    110
South America      34
Asia               10
Name: count, dtype: int64


Para ordenar os artistas por continente, primeiro mapeei cada país de origem dos artistas a seu respectivo continente usando um dicionário country_to_continent. Este mapeamento incluiu uma ampla gama de países distribuídos por todos os continentes habitados: América do Norte, América do Sul, Europa, Ásia, Oceania e África. Com esse mapeamento, atualizei o dataframe dos artistas, adicionando uma nova coluna que indicava o continente de cada artista com base em seu país de origem.

Depois de estabelecer o continente para cada artista, filtrei os dados para excluir artistas de países não mapeados ou categorizados como 'Desconhecido'. Com esse conjunto de dados filtrado, realizei uma contagem de quantos artistas havia por continente usando a função value_counts() no dataframe. Esta função automaticamente agrupa e conta o número de ocorrências (neste caso, artistas) para cada continente, fornecendo uma visão clara de quantos artistas estão representados em cada continente na base de dados.

A ordenação final dos artistas por continente refletiu o número de artistas associados a cada continente, permitindo identificar quais continentes têm a maior representação de artistas na base de dados.

### B. estados.csv

 Qual é a dimensão do conjunto de dados?

In [14]:
# Carregando o arquivo csv
estados_df = pd.read_csv('/home/hjhgjhgjhg/Downloads/csv2/estados.csv')

# Obtendo as dimensões do arquivo
dimensions = estados_df.shape
print("Dimensao do conjunto de dados do csv de estados:", dimensions)

# Imprimindo as dimensões do arquivo
print(f"Ou seja: {dimensions[0]} linhas e {dimensions[1]} colunas")

Dimensao do conjunto de dados do csv de estados: (550, 2)
Ou seja: 550 linhas e 2 colunas


Existe valores NaN?


In [15]:
print("\nQuantidade de valores NaN por coluna:")

# Contando a quantidade de valores NaN em cada coluna
nan_count = estados_df.isna().sum()

# Imprimindo a quantidade de valores NaN em cada coluna
for i in range(len(nan_count)):
    print(f"A coluna '{nan_count.index[i]}' contém {nan_count[i]} valores NaN")


Quantidade de valores NaN por coluna:
A coluna 'Estado' contém 0 valores NaN
A coluna 'Municipio' contém 0 valores NaN


/tmp/ipykernel_48438/2802910142.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"A coluna '{nan_count.index[i]}' contém {nan_count[i]} valores NaN")


Não existem valores não numéricos nesse arquivo csv em nenhuma das colunas.

### Qual é o valor mais comum de número de canções criadas em colaboração entre dois artistas

In [9]:
moda_collab_songs = artists_df['collab_songs'].mode()[0]

print(f"O valor mais comum de número de canções criadas em colaboração entre dois artistas é: {moda_collab_songs}")

O valor mais comum de número de canções criadas em colaboração entre dois artistas é: 10


### SEGUNDA PARTE

### Globalmente, os artistas têm mais tendência em colaborar com artistas do mesmo país, ou de
### países diferentes?

In [12]:
# Carregando o arquivo de colaborações para analisar as tendências de colaboração dos artistas.
collaborations_df = pd.read_csv('/home/hjhgjhgjhg/Downloads/csv/collaborations.csv')


# Mapeando o país de origem para cada artista em 'collaborations_df_new'
artist_to_country_map_new = artists_df.set_index('artist')['country'].to_dict()

# Aplicando o mapeamento para obter o país de origem de artist1 e artist2
collaborations_df['country1'] = collaborations_df['artist1'].map(artist_to_country_map_new)
collaborations_df['country2'] = collaborations_df['artist2'].map(artist_to_country_map_new)

# Determinando se a colaboração é nacional ou internacional
collaborations_df['collab_type'] = collaborations_df.apply(lambda x: 'National' if x['country1'] == x['country2'] else 'International', axis=1)

# Calculando a proporção global de tipos de colaboração
global_collab_trends_new = collaborations_df['collab_type'].value_counts(normalize=True) * 100

# Identificando as nacionalidades que mais fazem colaborações internacionais e nacionais
top_international_collab_countries_new = collaborations_df[collaborations_df['collab_type'] == 'International']['country1'].value_counts().head()
top_national_collab_countries_new = collaborations_df[collaborations_df['collab_type'] == 'National']['country1'].value_counts().head()

global_collab_trends_new, top_international_collab_countries_new, top_national_collab_countries_new


(collab_type
 National         54.723926
 International    45.276074
 Name: proportion, dtype: float64,
 country1
 United States      314
 Puerto Rico        181
 United Kingdom     150
 Colombia           108
 Other countries     90
 Name: count, dtype: int64,
 country1
 United States     921
 Puerto Rico       206
 Brazil             54
 United Kingdom     31
 Colombia           29
 Name: count, dtype: int64)

Globalmente, os artistas têm mais tendência a colaborar com artistas do mesmo país, conforme indicado pela proporção de 54.72% de colaborações nacionais em comparação com 45.28% de colaborações internacionais.

Os artistas dos Estados Unidos são os que mais privilegiam as colaborações internacionais, seguidos por artistas de Porto Rico, Reino Unido e Colômbia. Para colaborações nacionais, novamente os artistas dos Estados Unidos lideram, com Porto Rico e Brasil também mostrando forte preferência por colaborar dentro do seu próprio país.